src/flb_output.c

```diff
@@ -173,6 +173,98 @@ void flb_output_coro_add(struct flb_output_instance *ins, struct flb_coro *coro)
     mk_list_add(&out_flush->_head, &ins->flush_list);
 }
 
+/*
+ * Queue a task to be flushed at a later time
+ */
+static int flb_output_task_queue_enqueue(struct flb_task_queue *queue,
+                                         struct flb_task *task,
+                                         struct flb_output_instance *out_ins,
+                                         struct flb_config *config)
+{
+    struct flb_task_enqueued *queued_task;
+
+    queued_task = flb_malloc(sizeof(struct flb_task_enqueued));
+    if (!queued_task) {
+        flb_errno();
+        return -1;
+    }
+    queued_task->out_instance = out_ins;
+    queued_task->task = task;
+    queued_task->config = config;
+
+    mk_list_add(&queued_task->_head, &queue->pending);
+    return 0;
+}
+
+/*
+ * Pop task from pending queue and flush it
+ */
+static int flb_output_task_queue_flush_one(struct flb_task_queue *queue)
+{
+    struct flb_task_enqueued *queued_task;
+    int ret;
+
+    ret = mk_list_is_empty(&queue->pending);
+    if (ret == 0) {
+        flb_error("Attempting to pop task queue with no tasks enqueued");
+        return -1;
+    }
+
+    queued_task = mk_list_entry_first(&queue->pending, struct flb_task_enqueued, _head);
+    mk_list_del(&queued_task->_head);
+    mk_list_add(&queued_task->_head, &queue->in_progress);
+    ret = flb_output_task_flush(queued_task->task,
+                                queued_task->out_instance,
+                                queued_task->config);
+
+    return ret;
+}
+
+/*
+ * Will either run or queue running a single task
+ */
+int flb_output_task_singleplex_enqueue(struct flb_task_queue *queue,
```

Comment:
nit: I noticed this function has a return code, which in one case can be -1 if there was simply no pending task, but then everywhere this is called the return code is never checked... is this all on purpose? Do we need a return code then?

Reply:
I'm just following the convention set forth by flb_output_task_flush. It has a return value that isn't used too. I think as long as the function returning an error doesn't put fluent bit into a bad state, then it is OK.

So if there is an error nothing gets added to the pending queue. That means that the next flush will still be able to complete successfully

https://github.com/fluent/fluent-bit/blob/356a11b63ee4cae5d816aecac8814276cf57ea11/src/flb_output.c#L272